File: sdk/php/tests/robustness/ws/ReliabilityTest.php
<?php


use KuCoin\UniversalSDK\Api\DefaultClient;
use KuCoin\UniversalSDK\Common\Logger;
use KuCoin\UniversalSDK\Generate\Futures\FuturesPublic\TickerV1Event;
use KuCoin\UniversalSDK\Generate\Spot\Market\GetAllSymbolsReq;
use KuCoin\UniversalSDK\Generate\Spot\SpotPublic\AllTickersEvent;
use KuCoin\UniversalSDK\Generate\Spot\SpotPublic\TickerEvent;
use KuCoin\UniversalSDK\Generate\Spot\SpotPublic\TradeEvent;
use KuCoin\UniversalSDK\Model\ClientOptionBuilder;
use KuCoin\UniversalSDK\Model\Constants;
use KuCoin\UniversalSDK\Model\TransportOptionBuilder;
use KuCoin\UniversalSDK\Model\WebSocketClientOptionBuilder;
use PHPUnit\Framework\TestCase;
use React\EventLoop\Loop;
use React\Promise\Deferred;
use React\Promise\PromiseInterface;
use function React\Promise\all;


function waitFor(float $seconds, $result): PromiseInterface
{
    $deferred = new Deferred();

    Loop::get()->addTimer($seconds, function () use ($result, $deferred) {
        $deferred->resolve($result);
    });

    return $deferred->promise();
}

class ReliabilityTest extends TestCase
{
    public function testEventCallback()
    {
        $key = getenv('API_KEY') ?: '';
        $secret = getenv('API_SECRET') ?: '';
        $passphrase = getenv('API_PASSPHRASE') ?: '';

        $httpTransportOption = (new TransportOptionBuilder())
            ->setKeepAlive(true)
            ->setTotalTimeout(1)
            ->setRetryDelay(0)
            ->setMaxRetries(1)
            ->build();


        $eventCounter = 0;
        $websocketOption = (new WebSocketClientOptionBuilder())->setEventCallback(function (string $eventType, string $eventMessage) use (&$eventCounter) {
            Logger::info("event called", ["eventType" => $eventType, "eventMessage" => $eventMessage]);
            $eventCounter++;
        })->build();

        // Create a client using the specified options
        $clientOption = (new ClientOptionBuilder())
            ->setKey($key)
            ->setSecret($secret)
            ->setPassphrase($passphrase)
            ->setSpotEndpoint(Constants::GLOBAL_API_ENDPOINT)
            ->setFuturesEndpoint(Constants::GLOBAL_FUTURES_API_ENDPOINT)
            ->setBrokerEndpoint(Constants::GLOBAL_BROKER_API_ENDPOINT)
            ->setTransportOption($httpTransportOption)
            ->setWebSocketClientOption($websocketOption)
            ->build();

        // Create or get the global event loop
        $loop = Loop::get();

        $client = new DefaultClient($clientOption, $loop);
        $spotWs = $client->wsService()->newSpotPublicWS();

        // Start connection
        $spotWs->start()->then(function () use ($spotWs, $loop) {
            Logger::info("WebSocket started");

            // Subscribe to allTickers
            return $spotWs->allTickers(
            // Called when data is received
                function (string $topic, string $subject, AllTickersEvent $data) {

                },
                // Called when subscription is successful
                function (string $id) use ($spotWs, $loop) {
                    Logger::info("Subscribed with ID: $id");

                    // Schedule unsubscribe and shutdown after 5 seconds
                    $loop->addTimer(5, function () use ($id, $spotWs) {
                        Logger::info("Unsubscribing...");
                        $spotWs->unSubscribe($id)->finally(function () use ($spotWs) {
                            $spotWs->stop();
                        });
                    });
                },
                // Called when subscription fails
                function (Exception $e) use ($spotWs) {
                    Logger::error("Subscription failed", ['error' => $e->getMessage()]);
                    $spotWs->stop();
                }
            );
        })->catch(function (Exception $e) use ($spotWs) {
            self::fail($e->getMessage());
            $spotWs->stop();
        });

        // Run the event loop to process async tasks
        $loop->run();

        self::assertTrue($eventCounter > 0);
    }


    public function getClient()
    {
        $key = getenv('API_KEY') ?: '';
        $secret = getenv('API_SECRET') ?: '';
        $passphrase = getenv('API_PASSPHRASE') ?: '';

        $httpTransportOption = (new TransportOptionBuilder())
            ->setKeepAlive(true)
            ->setTotalTimeout(1)
            ->setRetryDelay(0)
            ->setMaxRetries(1)
            ->build();


        $websocketOption = (new WebSocketClientOptionBuilder())->setEventCallback(function (string $eventType, string $eventMessage) use (&$eventCounter) {
            Logger::info("event called", ["eventType" => $eventType, "eventMessage" => $eventMessage]);
        })->build();

        // Create a client using the specified options
        $clientOption = (new ClientOptionBuilder())
            ->setKey($key)
            ->setSecret($secret)
            ->setPassphrase($passphrase)
            ->setSpotEndpoint(Constants::GLOBAL_API_ENDPOINT)
            ->setFuturesEndpoint(Constants::GLOBAL_FUTURES_API_ENDPOINT)
            ->setBrokerEndpoint(Constants::GLOBAL_BROKER_API_ENDPOINT)
            ->setTransportOption($httpTransportOption)
            ->setWebSocketClientOption($websocketOption)
            ->build();

        $loop = Loop::get();

        return new DefaultClient($clientOption, $loop);
    }

    public function testReconnect()
    {
        $loop = Loop::get();

        $spotWs = self::getClient()->wsService()->newSpotPublicWS();

        // Start connection
        $spotWs->start()->then(function () use ($spotWs, $loop) {
            Logger::info("WebSocket started");

            // Subscribe to allTickers
            return $spotWs->allTickers(
            // Called when data is received
                function (string $topic, string $subject, AllTickersEvent $data) {
                    // pass
                },
                null,
                // Called when subscription fails
                null,
            );
        })->then(function () {
            return waitFor(3600, []);
        })->catch(function (Exception $e) {
            self::fail($e->getMessage());
        })->finally(function () use ($spotWs, $loop) {
            $spotWs->stop();
        });

        // Run the event loop to process async tasks
        $loop->run();
    }

    public function testReconnect2()
    {
        $loop = Loop::get();

        $client = self::getClient();
        $spotWs = $client->wsService()->newSpotPublicWS();

        // Start connection
        $spotWs->start()->then(function () use ($client, $spotWs, $loop) {
            Logger::info("WebSocket started");

            // Subscribe to allTickers
            $symbolsResponse = $client->restService()->getSpotService()->
            getMarketApi()->getAllSymbols(GetAllSymbolsReq::create(["market" => "USDS"]));

            $symbols = [];
            foreach ($symbolsResponse->data as $symbol) {
                $symbols[] = $symbol->symbol;
            }

            if (count($symbols) > 100) {
                $symbols = array_slice($symbols, 0, 100);
            }

            $promise = [];

            foreach ($symbols as $symbol) {
                $promise[] = $spotWs->ticker([$symbol], function (string $topic, string $subject, TickerEvent $data) {
                    // pass
                }, function ($id) {
                    Logger::info("subscribed with ID: $id");
                }, null);
            }

            return all($promise);
        })->then(function () {
            return waitFor(3600, []);
        })->catch(function (Exception $e) {
            self::fail($e->getMessage());
        })->finally(function () use ($spotWs, $loop) {
            $spotWs->stop();
        });

        // Run the event loop to process async tasks
        $loop->run();
    }


    public function testReconnect3()
    {
        $loop = Loop::get();

        $client = self::getClient();
        $spotWs = $client->wsService()->newSpotPublicWS();

        // Start connection
        $spotWs->start()->then(function () use ($client, $spotWs, $loop) {
            Logger::info("WebSocket started");

            $promise = [];

            $promise[] = $spotWs->ticker(["BTC-USDT"], function (string $topic, string $subject, TickerEvent $data) {
                // pass
            }, function ($id) {
                Logger::info("subscribed with ID: $id");
            }, null);

            $promise[] = $spotWs->trade(["BTC-USDT"], function (string $topic, string $subject, TradeEvent $data) {
                // pass
            }, function ($id) {
                Logger::info("subscribed with ID: $id");
            }, null);

            return all($promise);
        })->then(function () {
            return waitFor(3600, []);
        })->catch(function (Exception $e) {
            self::fail($e->getMessage());
        })->finally(function () use ($spotWs, $loop) {
            $spotWs->stop();
        });

        // Run the event loop to process async tasks
        $loop->run();
    }


    public function testReconnect4()
    {
        $loop = Loop::get();

        $client = self::getClient();
        $spotWs = $client->wsService()->newSpotPublicWS();
        $futuresWs = $client->wsService()->newFuturesPublicWS();
        // Start connection
        $spotWs->start()->then(function () use ($client, $spotWs, $loop) {
            Logger::info("WebSocket started");

            return $spotWs->ticker(["BTC-USDT"], function (string $topic, string $subject, TickerEvent $data) {
                // pass
            }, function ($id) {
                Logger::info("subscribed with ID: $id");
            }, null);
        })->then(function () {
            return waitFor(3600, []);
        })->catch(function (Exception $e) {
            self::fail($e->getMessage());
        })->finally(function () use ($spotWs, $loop) {
            $spotWs->stop();
        });

        $futuresWs->start()->then(function () use ($client, $futuresWs, $loop) {
            Logger::info("WebSocket started");

            return $futuresWs->tickerV1("XBTUSDTM", function (string $topic, string $subject, TickerV1Event $data) {
                // pass
            }, function ($id) {
                Logger::info("subscribed with ID: $id");
            }, null);
        })->then(function () {
            return waitFor(3600, []);
        })->catch(function (Exception $e) {
            self::fail($e->getMessage());
        })->finally(function () use ($futuresWs, $loop) {
            $futuresWs->stop();
        });

        // Run the event loop to process async tasks
        $loop->run();

    }
}

File: sdk/java/src/test/java/com/kucoin/universal/sdk/test/robustness/ws/ReliabilityTest.java
package com.kucoin.universal.sdk.test.robustness.ws;

import com.kucoin.universal.sdk.api.DefaultKucoinClient;
import com.kucoin.universal.sdk.api.KucoinClient;
import com.kucoin.universal.sdk.generate.spot.market.GetAllSymbolsData;
import com.kucoin.universal.sdk.generate.spot.market.GetAllSymbolsReq;
import com.kucoin.universal.sdk.generate.spot.market.GetAllSymbolsResp;
import com.kucoin.universal.sdk.generate.spot.spotpublic.AllTickersEvent;
import com.kucoin.universal.sdk.generate.spot.spotpublic.SpotPublicWs;
import com.kucoin.universal.sdk.generate.spot.spotpublic.TradeEvent;
import com.kucoin.universal.sdk.model.ClientOption;
import com.kucoin.universal.sdk.model.Constants;
import com.kucoin.universal.sdk.model.TransportOption;
import com.kucoin.universal.sdk.model.WebSocketClientOption;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

@Slf4j
public class ReliabilityTest {

  @Test
  public void testCallback() {
    String key = System.getenv("API_KEY");
    String secret = System.getenv("API_SECRET");
    String passphrase = System.getenv("API_PASSPHRASE");

    AtomicInteger atomicInteger = new AtomicInteger();

    ClientOption clientOpt =
        ClientOption.builder()
            .key(key)
            .secret(secret)
            .passphrase(passphrase)
            .spotEndpoint(Constants.GLOBAL_API_ENDPOINT)
            .futuresEndpoint(Constants.GLOBAL_FUTURES_API_ENDPOINT)
            .brokerEndpoint(Constants.GLOBAL_BROKER_API_ENDPOINT)
            .transportOption(TransportOption.defaults())
            .websocketClientOption(
                WebSocketClientOption.builder()
                    .eventCallback(
                        ((event, message) -> {
                          atomicInteger.incrementAndGet();
                          log.info("Event: {}, {}", event, message);
                        }))
                    .build())
            .build();

    KucoinClient kucoinClient = new DefaultKucoinClient(clientOpt);
    SpotPublicWs spotPublicWs = kucoinClient.getWsService().newSpotPublicWS();
    spotPublicWs.start();
    spotPublicWs.stop();

    Assertions.assertEquals(3, atomicInteger.get());
  }

  @Test
  public void testReconnect() throws Exception {
    String key = System.getenv("API_KEY");
    String secret = System.getenv("API_SECRET");
    String passphrase = System.getenv("API_PASSPHRASE");

    AtomicInteger atomicInteger = new AtomicInteger();

    ClientOption clientOpt =
        ClientOption.builder()
            .key(key)
            .secret(secret)
            .passphrase(passphrase)
            .spotEndpoint(Constants.GLOBAL_API_ENDPOINT)
            .futuresEndpoint(Constants.GLOBAL_FUTURES_API_ENDPOINT)
            .brokerEndpoint(Constants.GLOBAL_BROKER_API_ENDPOINT)
            .transportOption(TransportOption.defaults())
            .websocketClientOption(
                WebSocketClientOption.builder()
                    .eventCallback(
                        ((event, message) -> {
                          atomicInteger.incrementAndGet();
                          log.info("Event: {}, {}", event, message);
                        }))
                    .build())
            .build();

    KucoinClient kucoinClient = new DefaultKucoinClient(clientOpt);
    SpotPublicWs spotPublicWs = kucoinClient.getWsService().newSpotPublicWS();
    spotPublicWs.start();

    spotPublicWs.allTickers((String topic, String subject, AllTickersEvent data) -> {});

    Thread.sleep(1000 * 12000);
    spotPublicWs.stop();
  }

  private static void noop(String topic, String subject, TradeEvent data) {}

  @Test
  public void testReconnect1() throws Exception {
    String key = System.getenv("API_KEY");
    String secret = System.getenv("API_SECRET");
    String passphrase = System.getenv("API_PASSPHRASE");

    AtomicInteger atomicInteger = new AtomicInteger();

    ClientOption clientOpt =
        ClientOption.builder()
            .key(key)
            .secret(secret)
            .passphrase(passphrase)
            .spotEndpoint(Constants.GLOBAL_API_ENDPOINT)
            .futuresEndpoint(Constants.GLOBAL_FUTURES_API_ENDPOINT)
            .brokerEndpoint(Constants.GLOBAL_BROKER_API_ENDPOINT)
            .transportOption(TransportOption.defaults())
            .websocketClientOption(
                WebSocketClientOption.builder()
                    .eventCallback(
                        ((event, message) -> {
                          atomicInteger.incrementAndGet();
                          log.info("Event: {}, {}", event, message);
                        }))
                    .build())
            .build();

    KucoinClient kucoinClient = new DefaultKucoinClient(clientOpt);

    GetAllSymbolsResp resp =
        kucoinClient
            .getRestService()
            .getSpotService()
            .getMarketApi()
            .getAllSymbols(GetAllSymbolsReq.builder().market("USDS").build());
    List<GetAllSymbolsData> data = resp.getData();
    if (data.size() > 100) {
      data = data.subList(0, 100);
    }

    SpotPublicWs spotPublicWs = kucoinClient.getWsService().newSpotPublicWS();
    spotPublicWs.start();

    data.forEach(
        d -> {
          spotPublicWs.trade(new String[] {d.getSymbol()}, ReliabilityTest::noop);
        });

    Thread.sleep(1000 * 12000);
    spotPublicWs.stop();
  }

  @Test
  public void testReconnect2() throws Exception {
    String key = System.getenv("API_KEY");
    String secret = System.getenv("API_SECRET");
    String passphrase = System.getenv("API_PASSPHRASE");

    AtomicInteger atomicInteger = new AtomicInteger();

    ClientOption clientOpt =
        ClientOption.builder()
            .key(key)
            .secret(secret)
            .passphrase(passphrase)
            .spotEndpoint(Constants.GLOBAL_API_ENDPOINT)
            .futuresEndpoint(Constants.GLOBAL_FUTURES_API_ENDPOINT)
            .brokerEndpoint(Constants.GLOBAL_BROKER_API_ENDPOINT)
            .transportOption(TransportOption.defaults())
            .websocketClientOption(
                WebSocketClientOption.builder()
                    .eventCallback(
                        ((event, message) -> {
                          atomicInteger.incrementAndGet();
                          log.info("Event: {}, {}", event, message);
                        }))
                    .build())
            .build();

    KucoinClient kucoinClient = new DefaultKucoinClient(clientOpt);

    SpotPublicWs spotPublicWs = kucoinClient.getWsService().newSpotPublicWS();
    spotPublicWs.start();

    spotPublicWs.trade(new String[] {"BTC-USDT", "ETH-USDT"}, ReliabilityTest::noop);

    Thread.sleep(1000 * 12000);
    spotPublicWs.stop();
  }
}
